### ISSUE TITLE
SDL2 compiling

### ISSUE THREAD

**Issue by igroglaz** [draconisPW/PWMAngband#503]
To this ticket I'll send stuff which were discussed in our discord. Please continue discussion there, so data won't be lost 🎅 
----------------------

[19:25] Serega88: @PowerWyrm  Tried to compile SDL2 - Borland C++ 6
downloaded. Development Libraries Windows (Visual C ++ 32/64-bit) \lib\x86 , all in archive 'PWMAngband-sdl2.zip'
displays errors in console
Warning W8027 client\..\_SDL2\SDL_stdinc.h 497: Functions containing switch are not expanded inline
Error E2121 client\..\_SDL2\SDL_endian.h 116: Function call missing ) in function SDL_Swap16
Error E2121 client\..\_SDL2\SDL_endian.h 164: Function call missing ) in function SDL_Swap32
Warning W8027 client\..\_SDL2\SDL_endian.h 215: Functions containing local classes are not expanded inline
Error E2121 client\..\_SDL2\SDL_endian.h 223: Function call missing ) in function SDL_Swap64
Warning W8027 client\..\_SDL2\SDL_endian.h 263: Functions containing local classes are not expanded inline

possibly make a fix for SDL_stdinc.h, SDL_endian.h
and there are many errors in main-sdl2.c 'Declaration is not allowed here in function'
https://cboard.cprogramming.com/c-programming/118264-why-do-i-get-error.html
Error E2140  Declaration is not allowed here in function ...
GL.Sam - "Preferable declare an item at the start of function block.
C99, hovewer, is less strict."

gcc uses CFLAGS  -std=c99 is it possible to enable c99 in bcc for makefile.sdl2 ?

https://en.wikipedia.org/wiki/List_of_compilers#C_compilers
C++Builder  Embarcadero  C99 - Yes

==========================
[PWMAngband-sdl2.zip](https://github.com/draconisPW/PWMAngband/files/7777484/PWMAngband-sdl2.zip)
[Readme-sdl2.txt](https://github.com/draconisPW/PWMAngband/files/7777485/Readme-sdl2.txt)

**Comment by igroglaz** [comment]
[21:22] PowerWyrm: SDL2 doesn't compile with C++ Builder 6, the libs shipped with it aren't compatible
[21:23] PowerWyrm: (SDL2 uses directX version that is bigger than directX files in /vcl) 
[21:23] PowerWyrm: only SDL 1 libs are compatible
[21:24] PowerWyrm: you have to compile using recent VC++
[21:26] PowerWyrm: (I tried compiling and spent days removing the "declaration not allowed" errors + "keyword is reserved" errors because there were a lot of variables called "this" or "class" and I managed to remove all compile errors but after that there were unsolvable linker errors due to wrong version of directX)

**Comment by igroglaz** [comment]
[17:50] Serega88: SDL2 should not depend on directX (maybe these are directX modules)
in sdl2 client \lib\user\sdl2init.txt  window-renderer:0:hardware
SDL_RENDERER_ACCELERATED) ? "hardware" : "software"); https://wiki.libsdl.org/SDL_RendererFlags
for example, I want to enable OpenGL https://lazyfoo.net/tutorials/SDL/51_SDL_and_modern_opengl/index.php
include <gl\glew.h>  include <SDL_opengl.h>

found https://github.com/libsdl-org/SDL/issues/1673  "We can add a compiler preprocessor check for Embarcadero C++" it should compile by Embarcadero C++
https://github.com/libsdl-org/SDL/commit/f5e636a62169f3429f2d03c237840bdc3020305f
I downloaded Embarcadero C++ 7.30 for Win32 https://altd.embarcadero.com/download/bcppbuilder/BCC102.zip
after compilation 'PWMAngband-master' is displayed SDL2.LIB' contains invalid OMF record, type 0x21 (possibly COFF)
found  https://docwiki.embarcadero.com/RADStudio/Sydney/en/IMPLIB.EXE,_the_Import_Library_Tool_for_Win32
http://gnuwin32.sourceforge.net/msimport.html
IMPLIB takes as input either DLLs or module definition files, or both, and produces an import library (.lib) as output. 
and created .lib files from 'VC-x86' folder (C:\BCC102\bin>implib.exe -a -c SDL2_image.lib SDL2_image.dll), details in archive Readme.txt
compiles fine. mangclient_sdl2.exe 🙂
I solved the problem with the mouse by replacing the libraries from 'fix-noteye-84', maybe some libraries are superfluous, maybe not.
maybe need version 'noteye-84' SDL 2.0.3.0 ,  .h files and convert dll-lib implib.exe -a -c SDL2.lib SDL2.dll ...
I don't know what to do with SDL2main.lib
please check. all files in archive 🎄 

![pwma-sdl2](https://user-images.githubusercontent.com/39904162/147418522-8178d1c2-14e2-419d-b1e7-f285af3314a1.png)
[pwma-exe-sdl2.zip](https://github.com/draconisPW/PWMAngband/files/7777486/pwma-exe-sdl2.zip)
[pwma-sdl2.zip](https://github.com/draconisPW/PWMAngband/files/7777487/pwma-sdl2.zip)

**Comment by s88100** [comment]
problem with mouse Windows 10 is solved by copying with replacement SDL2-devel-2.0.16\lib\x86\SDL2.dll
https://www.libsdl.org/release/SDL2-devel-2.0.16-VC.zip
and new libraries from pwma-sdl2.zip \VC-x86
_downloaded dll, .h files, version SDL 2.0.3.0 2014 noteye-84 and converted implib.exe -a -c '.lib' files,  doesn't compile, 
only need new 2021 development libraries._

**Comment by s88100** [comment]
in [pwma-sdl2.zip](https://github.com/draconisPW/PWMAngband/files/7777487/pwma-sdl2.zip) PWMAngband-master/src/bcc32c/client-sdl2.bat
need to fix  `del ..\client\c-cmd.obj`  ...
and makefile BCCDIR = "C:\Data\extra\PWMAngband workspace\BCC102"

> error: no such file or directory: 'workspace\BCC102\bin\bcc32c.exe'

I placed in BCCDIR = "C:\BCC102"

**Comment by s88100** [comment]
src/bcc32c/client-sdl2.bat
there are two  warnings in terminal
```
..\common\z-form.c:476:38: warning: second argument to 'va_arg' is of promotable type 'short'; this va_arg has undefined
      behavior because arguments will be promoted to 'int'
                    arg = va_arg(vp, short);
                                     ^~~~~
C:\BCC102\bin\..\include\windows\crtl\stdarg.h:40:50: note: expanded from macro 'va_arg'
#define va_arg(ap, type)    __builtin_va_arg(ap, type)
                                                 ^
..\common\z-form.c:524:38: warning: second argument to 'va_arg' is of promotable type 'unsigned short'; this va_arg has
      undefined behavior because arguments will be promoted to 'int'
                    arg = va_arg(vp, unsigned short);
                                     ^~~~~~~~
C:\BCC102\bin\..\include\windows\crtl\stdarg.h:40:50: note: expanded from macro 'va_arg'
#define va_arg(ap, type)    __builtin_va_arg(ap, type)
                                                 ^
2 warnings generated.
```
to read file mangclient.ini
https://github.com/draconisPW/PWMAngband/blob/master/src/client/conf.c#L61
`if (suffix(path, "_sdl2")) path[strlen(path) - 5] = '\0';`

**Comment by s88100** [comment]
need library SDL2_image-2.0.5\lib\x86\libjpeg-9.dll
https://www.libsdl.org/projects/SDL_image/release/SDL2_image-devel-2.0.5-VC.zip
to open wallpaper lib\user\sdl2init.txt
`window-wallpaper-path:0:.\lib\icons\01.jpg`
I don’t know about .tiff format > libtiff-5.dll

**Comment by s88100** [comment]
_for correct walking/running through shift+ Numlock keyboard, possibly enable it by default._
/src/client/main-sdl2.c :L559
```
/* True if KC_MOD_KEYPAD will be sent for numeric keypad keys at the expense
 * of not handling some keyboard layouts properly. */
static int g_kp_as_mod = 1;
```

**Comment by s88100** [comment]
in GNU/Linux sdl1 client PWMAngband-master/mk/buildsys.mk

> CFLAGS = -g -O2 -DHAVE_CONFIG_H -W -Wall -Wextra -Wno-unused-parameter -pedantic -Wno-missing-field-initializers -I/usr/include/SDL -D_GNU_SOURCE=1 -D_REENTRANT ${SANITIZE_FLAGS}
> LIBS =  -L/usr/lib/x86_64-linux-gnu -lSDL -lSDL_image -lSDL_ttf -lSDL_mixer -lm

in sdl2 client PWMAngband-master/mk/buildsys.mk

> CFLAGS = -g -O2 -DHAVE_CONFIG_H -W -Wall -Wextra -Wno-unused-parameter -pedantic -Wno-missing-field-initializers -I/usr/include/SDL2 -D_REENTRANT ${SANITIZE_FLAGS}
> LIBS =  -lSDL2 -lSDL2_image -lSDL2_ttf -lSDL2_mixer -lm

build system is not defined here "-D_GNU_SOURCE=1"

if put a flag it will work fine
```
./autogen.sh
./configure --disable-curses --enable-sdl2 CFLAGS="-D_GNU_SOURCE=1"
make
make install
```
then it is required to remove workaround in /src/client/ui-init.c , /src/server/init.c
https://github.com/draconisPW/PWMAngband/blob/master/src/client/ui-init.c#L163
https://github.com/draconisPW/PWMAngband/blob/master/src/server/init.c#L409
```
void create_needed_dirs(void) ...
#if !defined(USE_SDL2)
```
and there is no warnings 'implicit declaration' in the terminal and the music is opened from the folder.
maybe need to take something from build system V 4.2.3

**Comment by draconisPW** [comment]
Fixed by MR #504.
